ref/gl/vgl_shim/vgl_shim.c

```diff
@@ -0,0 +1,572 @@
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+*/
+
+/*
+	this is a "replacement" for vitaGL's immediate mode tailored specifically for xash
+	this will only provide performance gains if vitaGL is built with DRAW_SPEEDHACK=1
+	since that makes it assume that all vertex data pointers are GPU-mapped
+*/
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <malloc.h>
+#include <vitaGL.h>
+
+#include "port.h"
+#include "xash3d_types.h"
+#include "cvardef.h"
+#include "const.h"
+#include "com_model.h"
+#include "cl_entity.h"
+#include "render_api.h"
+#include "protocol.h"
+#include "dlight.h"
+#include "ref_api.h"
+#include "com_strings.h"
+#include "vgl_shim.h"
+
+#define MAX_SHADERLEN 4096
+// increase this when adding more attributes
+#define MAX_PROGS 32
+
+extern ref_api_t gEngfuncs;
+
+enum vgl_attrib_e
+{
+	VGL_ATTR_POS       = 0, // 1
+	VGL_ATTR_COLOR     = 1, // 2
+	VGL_ATTR_TEXCOORD0 = 2, // 4
+	VGL_ATTR_TEXCOORD1 = 3, // 8
+	VGL_ATTR_MAX
+};
+
+// continuation of previous enum
+enum vgl_flag_e
+{
+	VGL_FLAG_ALPHA_TEST = VGL_ATTR_MAX, // 16
+	VGL_FLAG_FOG,                       // 32
+	VGL_FLAG_MAX
+};
+
+typedef struct
+{
+	GLuint flags;
+	GLint attridx[VGL_ATTR_MAX];
+	GLuint glprog;
+	GLint ucolor;
+	GLint ualpha;
+	GLint utex0;
+	GLint utex1;
+	GLint ufog;
+} vgl_prog_t;
+
+static const char *vgl_vert_src =
+#include "vgl_shaders/vertex.cg.inc"
+;
+
+static const char *vgl_frag_src =
+#include "vgl_shaders/fragment.cg.inc"
+;
+
+static int vgl_init = 0;
+
+static struct
+{
+	GLfloat *attrbuf[VGL_ATTR_MAX];
+	GLuint cur_flags;
+	GLint begin;
+	GLint end;
+	GLenum prim;
+	GLfloat color[4];
+	GLfloat fog[4]; // color + density
+	GLfloat alpharef;
+	vgl_prog_t progs[MAX_PROGS];
+	vgl_prog_t *cur_prog;
+	GLboolean uchanged;
+} vgl;
+
+static const int vgl_attr_size[VGL_ATTR_MAX] = { 3, 4, 2, 2 };
+
+static const char *vgl_flag_name[VGL_FLAG_MAX] =
+{
+	"ATTR_POSITION",
+	"ATTR_COLOR",
+	"ATTR_TEXCOORD0",
+	"ATTR_TEXCOORD1",
+	"FEAT_ALPHA_TEST",
+	"FEAT_FOG",
+};
+
+static const char *vgl_attr_name[VGL_ATTR_MAX] =
+{
+	"inPosition",
+	"inColor",
+	"inTexCoord0",
+	"inTexCoord1",
+};
+
+// HACK: borrow alpha test and fog flags from internal vitaGL state
+extern GLboolean alpha_test_state;
+extern GLboolean fogging;
+
+static GLuint VGL_GenerateShader( const vgl_prog_t *prog, GLenum type )
+{
+	char *shader, shader_buf[MAX_SHADERLEN + 1];
+	char tmp[256];
+	int i;
+	GLint status, len;
+	GLuint id;
+
+	shader = shader_buf;
+	shader[0] = '\n';
+	shader[1] = 0;
+
+	for ( i = 0; i < VGL_FLAG_MAX; ++i )
+	{
+		snprintf( tmp, sizeof( tmp ), "#define %s %d\n", vgl_flag_name[i], prog->flags & ( 1 << i ) );
+		strncat( shader, tmp, MAX_SHADERLEN );
```

Comment:
Why not use crtlib functions here?
https://github.com/FWGS/xash3d-fwgs/blob/master/public/crtlib.h#L70=

Reply:
Done.